independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java

```diff
@@ -0,0 +1,215 @@
+package io.quarkus.paths;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.jar.Manifest;
+
+/**
+ * While {@link ArchivePathTree} implementation is thread-safe, this implementation
+ * provides a way for multiple concurrent users to share the same instance of an open archive,
+ * instead of every user opening its own copy of it.
+ */
+class SharedArchivePathTree extends ArchivePathTree {
+
+    private static final String DISABLE_JAR_CACHE_PROPERTY = "quarkus.bootstrap.disable-jar-cache";
+
+    private static final boolean ENABLE_SHARING;
+
+    // It would probably be better to have a weak hash map based cache,
+    // however as the first iteration it'll already be more efficient than before
+    private static final Map<Path, SharedArchivePathTree> cache;
+
+    static {
+        ENABLE_SHARING = !Boolean.getBoolean(DISABLE_JAR_CACHE_PROPERTY);
+        cache = ENABLE_SHARING ? new ConcurrentHashMap<>() : Map.of();
+    }
+
+    /**
+     * Returns an instance of {@link ArchivePathTree} for the {@code path} either from a cache
+     * or creates a new instance for it and puts it in the cache.
+     *
+     * @param path path to an archive
+     * @return instance of {@link ArchivePathTree}, never null
+     */
+    static ArchivePathTree forPath(Path path) {
+        return ENABLE_SHARING
+                ? cache.computeIfAbsent(path, k -> new SharedArchivePathTree(path))
```

Comment:
I think applying the following patch would make it more efficient and also will avoid the lambdas:

```diff
diff --git a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java
index 63c5a39b529..372dd1401d0 100644
--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java
@@ -40,9 +40,11 @@ class SharedArchivePathTree extends ArchivePathTree {
      * @return instance of {@link ArchivePathTree}, never null
      */
     static ArchivePathTree forPath(Path path) {
-        return ENABLE_SHARING
-                ? cache.computeIfAbsent(path, k -> new SharedArchivePathTree(path))
-                : new ArchivePathTree(path);
+        if (!ENABLE_SHARING) {
+            return new ArchivePathTree(path);
+        }
+
+        return cache.computeIfAbsent(path, SharedArchivePathTree::new);
     }
 
     /**
@@ -57,9 +59,11 @@ static ArchivePathTree forPath(Path path) {
      * @return instance of {@link ArchivePathTree}, never null
      */
     static ArchivePathTree forPath(Path path, PathFilter filter) {
-        return filter == null && ENABLE_SHARING
-                ? cache.computeIfAbsent(path, k -> new SharedArchivePathTree(path))
-                : new ArchivePathTree(path, filter);
+        if (filter != null) {
+            return new ArchivePathTree(path, filter);
+        }
+
+        return forPath(path);
     }
 
     /**
@@ -69,6 +73,10 @@ static ArchivePathTree forPath(Path path, PathFilter filter) {
      * @param path path to remove from the cache
      */
     static void removeFromCache(Path path) {
+        if (!ENABLE_SHARING) {
+            return;
+        }
+
         cache.remove(path);
     }
 

```

I can push to your branch if it makes sense to you.

Reply:
Looks good, thanks

Reply:
I'll apply the changes now, thanks.